File: api/app/Http/Controllers/Api/RegisterController.php
<?php
// app/Http/Controllers/Api/RegisterController.php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Plan;
use App\Models\SmsPreference;
use App\Models\Subscription;
use App\Models\Tenant;
use App\Models\User;
use App\Services\OtpService;
use App\Support\Phone;
use App\Support\TenantContext;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Hash;
use Illuminate\Support\Str;
use Illuminate\Validation\Rule;
use Illuminate\Validation\ValidationException;

class RegisterController extends Controller
{
    public function __construct(private OtpService $otp) {}

    /**
     * Yeni firma + yönetici kaydı. Deneme aboneliği başlatır.
     * Ardından OTP gönderir; doğrulama /auth/verify-otp ile yapılır.
     */
    public function register(Request $request): JsonResponse
    {
        $data = $request->validate([
            'name'         => ['required', 'string', 'max:255'],   // kişi adı
            'surname'      => ['nullable', 'string', 'max:255'],
            'company_name' => ['required', 'string', 'max:255'],
            'phone'        => ['required', 'string'],
            'email'        => ['nullable', 'email', 'max:255'],
            'password'     => ['required', 'string', 'min:6'],
            'plan'         => ['nullable', Rule::in(['trial', 'starter', 'pro', 'enterprise'])],
        ]);

        $phone = Phone::normalize($data['phone']);
        if (! Phone::isValid($phone)) {
            throw ValidationException::withMessages(['phone' => ['Geçerli bir cep telefonu girin.']]);
        }

        $planCode = $data['plan'] ?? 'trial';

        $result = DB::transaction(function () use ($data, $phone, $planCode) {
            $tenant = Tenant::create([
                'name'            => $data['company_name'],
                'slug'            => $this->uniqueSlug($data['company_name']),
                'phone'           => $phone,
                'email'           => $data['email'] ?? null,
                'plan'            => $planCode,
                'plan_expires_at' => now()->addDays(30),
                'sms_balance'     => 100,
                'is_active'       => true,
            ]);

            // tenant bağlamını set et ki HasTenant'lı kayıtlar otomatik tenant_id alsın
            TenantContext::set($tenant);

            $user = User::create([
                'tenant_id' => $tenant->id,
                'name'      => $data['name'],
                'surname'   => $data['surname'] ?? null,
                'phone'     => $phone,
                'email'     => $data['email'] ?? null,
                'password'  => Hash::make($data['password']),
                'role'      => User::ROLE_MANAGER,
                'is_active' => true,
            ]);

            $plan = Plan::where('code', $planCode)->first();
            Subscription::create([
                'tenant_id'          => $tenant->id,
                'plan_id'            => $plan?->id,
                'status'             => 'trialing',
                'started_at'         => now(),
                'current_period_end' => now()->addDays(30),
            ]);

            SmsPreference::create(['tenant_id' => $tenant->id]);

            return compact('tenant', 'user');
        });

        $devCode = $this->otp->generate($phone, $result['tenant']->id, 'login');

        return response()->json([
            'message'      => 'Firma kaydınız oluşturuldu. Doğrulama kodu gönderildi.',
            'requires_otp' => true,
            'phone'        => $phone,
            'tenant'       => [
                'id'   => $result['tenant']->id,
                'name' => $result['tenant']->name,
                'slug' => $result['tenant']->slug,
            ],
            'dev_code'     => config('app.debug') ? $devCode : null,
        ], 201);
    }

    private function uniqueSlug(string $name): string
    {
        $base = Str::slug($name) ?: 'firma';
        $slug = $base;
        $i = 1;
        while (Tenant::where('slug', $slug)->exists()) {
            $slug = $base . '-' . (++$i);
        }
        return $slug;
    }
}

File: api-net/Controllers/RegisterController.cs
using Liftonom.Api.Data;
using Liftonom.Api.Models;
using Liftonom.Api.Services;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace Liftonom.Api.Controllers;

[ApiController]
[Route("api/v1/auth")]
public class RegisterController(AppDbContext db, OtpService otp, ITenantContext tenantCtx, IHostEnvironment env) : ControllerBase
{
    public record RegisterDto(string Name, string? Surname, string CompanyName, string Phone, string? Email, string Password, string? Plan);

    [HttpPost("register")]
    public async Task<IActionResult> Register(RegisterDto dto)
    {
        var phone = PhoneHelper.Normalize(dto.Phone);
        if (!PhoneHelper.IsValid(phone)) throw new ApiException(422, "Geçerli bir cep telefonu girin.");
        var planCode = dto.Plan ?? "trial";
        var now = DateTime.UtcNow;

        await using var tx = await db.Database.BeginTransactionAsync();

        var tenant = new Tenant
        {
            Name = dto.CompanyName, Slug = await UniqueSlug(dto.CompanyName), Phone = phone, Email = dto.Email,
            Plan = planCode, PlanExpiresAt = now.AddDays(30), SmsBalance = 100, IsActive = true,
            CreatedAt = now, UpdatedAt = now,
        };
        db.Tenants.Add(tenant);
        await db.SaveChangesAsync();

        // Yeni kayıtlarda tenant bağlamını set et (sms_preferences vb. için)
        tenantCtx.TenantId = tenant.Id;

        var user = new User
        {
            TenantId = tenant.Id, Name = dto.Name, Surname = dto.Surname, Phone = phone, Email = dto.Email,
            Password = BCrypt.Net.BCrypt.HashPassword(dto.Password), Role = "manager", IsActive = true,
            CreatedAt = now, UpdatedAt = now,
        };
        db.Users.Add(user);

        var plan = await db.Plans.FirstOrDefaultAsync(p => p.Code == planCode);
        db.Subscriptions.Add(new Subscription
        {
            TenantId = tenant.Id, PlanId = plan?.Id, Status = "trialing", StartedAt = now,
            CurrentPeriodEnd = now.AddDays(30), CreatedAt = now, UpdatedAt = now,
        });
        db.SmsPreferences.Add(new SmsPreference { TenantId = tenant.Id, CreatedAt = now, UpdatedAt = now });

        await db.SaveChangesAsync();
        await tx.CommitAsync();

        var code = await otp.GenerateAsync(phone, tenant.Id, "login");

        return StatusCode(201, new
        {
            message = "Firma kaydınız oluşturuldu. Doğrulama kodu gönderildi.",
            requires_otp = true,
            phone,
            tenant = new { tenant.Id, tenant.Name, tenant.Slug },
            dev_code = env.IsDevelopment() ? code : null,
        });
    }

    private async Task<string> UniqueSlug(string name)
    {
        var baseSlug = Slugify(name);
        if (string.IsNullOrEmpty(baseSlug)) baseSlug = "firma";
        var slug = baseSlug;
        var i = 1;
        while (await db.Tenants.IgnoreQueryFilters().AnyAsync(t => t.Slug == slug))
            slug = $"{baseSlug}-{++i}";
        return slug;
    }

    private static string Slugify(string input)
    {
        var s = input.ToLowerInvariant().Trim();
        s = s.Replace('ı', 'i').Replace('ğ', 'g').Replace('ü', 'u').Replace('ş', 's').Replace('ö', 'o').Replace('ç', 'c');
        var chars = s.Select(c => char.IsLetterOrDigit(c) ? c : '-').ToArray();
        s = new string(chars);
        while (s.Contains("--")) s = s.Replace("--", "-");
        return s.Trim('-');
    }
}
